File: Packages/mcp-unity/Editor/Tools/ProfilerGetMemoryStatsTool.cs
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for getting Unity Profiler memory statistics.
    /// </summary>
    public class ProfilerGetMemoryStatsTool : McpToolBase
    {
        public ProfilerGetMemoryStatsTool()
        {
            Name = "profiler_get_memory_stats";
            Description = "Gets Unity Profiler memory statistics";
        }

        /// <summary>
        /// Execute the ProfilerGetMemoryStats tool with no parameters.
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject</param>
        public override JObject Execute(JObject parameters)
        {
            try
            {
                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = "Retrieved profiler memory statistics",
                    ["totalAllocatedMemoryMB"] = UnityEngine.Profiling.Profiler.GetTotalAllocatedMemoryLong() / 1048576.0,
                    ["totalReservedMemoryMB"] = UnityEngine.Profiling.Profiler.GetTotalReservedMemoryLong() / 1048576.0,
                    ["totalUnusedReservedMemoryMB"] = UnityEngine.Profiling.Profiler.GetTotalUnusedReservedMemoryLong() / 1048576.0,
                    ["monoUsedSizeMB"] = UnityEngine.Profiling.Profiler.GetMonoUsedSizeLong() / 1048576.0,
                    ["monoHeapSizeMB"] = UnityEngine.Profiling.Profiler.GetMonoHeapSizeLong() / 1048576.0
                };
            }
            catch (System.Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error getting profiler memory statistics: {ex.Message}",
                    "profiler_error"
                );
            }
        }
    }
}

File: Packages/mcp-unity/Server~/src/tools/profilerGetMemoryStatsTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';

const toolName = 'profiler_get_memory_stats';
const toolDescription = 'Gets Unity Profiler memory statistics';

const paramsSchema = z.object({});

export function registerProfilerGetMemoryStatsTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: z.infer<typeof paramsSchema>) {
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to get profiler memory statistics'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || 'Retrieved profiler memory statistics'
    }],
    data: {
      totalAllocatedMemoryMB: response.totalAllocatedMemoryMB,
      totalReservedMemoryMB: response.totalReservedMemoryMB,
      totalUnusedReservedMemoryMB: response.totalUnusedReservedMemoryMB,
      monoUsedSizeMB: response.monoUsedSizeMB,
      monoHeapSizeMB: response.monoHeapSizeMB,
      tempAllocatorSizeMB: response.tempAllocatorSizeMB
    }
  };
}
